scripts/format_code.py

```diff
@@ -0,0 +1,257 @@
+clang-format on them.  By default the tool will format the files in place.
+-noformat_file will run a check-only mode.
+
+Returns:
+   0: Either no files were found which need formatting, or if format_file was
+     enablled and the files were succesfully formatted.
+   1: Files which need formatting were found but weren't formatted
+     because -noformat_file was set.
+   2: The application wasn't configured properly and could not execute.
+"""
+import io
+import os
+import subprocess
+import sys
+
+from os import listdir
+from os.path import isfile, join
+from absl import app
+from absl import flags
+
+
+# Flag Definitions:
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_boolean("git_diff", False, "Use git-diff to assemble a file list")
+flags.DEFINE_string("git_range", "main..", "the range string when using "
+  "git-diff")
+flags.DEFINE_multi_string("file", None, "Append the filename to the list of "
+  "files to check.")
+flags.DEFINE_multi_string("d", None, 
+  "Append the directory to the file list to format.")
+flags.DEFINE_multi_string("dr", None, 
+  "Append the directory, and recurse through its children, and add their "
+  "contents to the list of files to check.")
+flags.DEFINE_boolean("format_file", True, "Format files in place")
+flags.DEFINE_boolean("verbose", False, "Execute in verbose mode")
+
+# Constants:
+
+# The list of file types to run clang-format on.   Used to filter out
+# results when searching across directories or git diffs.
+FILE_TYPE_EXTENSIONS = (".ccp", ".cc", ".c", ".h")
+
+# Common execution arguments across all clang-format executions.
+CLANG_FORMAT_BASE = "clang-format -style=file "
+
+# Flag which informs clang-format to alter the file in place, overwriting
+# the original file with thew newly formatted version.
+CLANG_FORMAT_PARM_IN_PLACE = "-i "
+
+# Flag which informs clang-format to use stdout to log the changes it would
+# make to a file.  Logging in XML format allows us to easily parse the results 
+# to detect if the file requires formatting changes.
+CLANG_FORMAT_PARAM_LOG_REPLACEMENTS = "-output-replacements-xml "
+
+# Conmmon exeuction arguments across all git diff executions.
+GIT_DIFF_BASE = "git diff "
+
+# Parameters to append for all git diff executions. The ACMRT filter lists
+# the files that have been (A)dded (C)opied (M)odified (R)enamed or (T)ype
+# changed.
+GIT_DIFF_PARAMS = " --name-only --diff-filter=ACMRT"
+
+# Functions:    
+
+def does_file_need_formatting(filename):
+  """Executes clang-format on the file to determine if it includes any
+  formatting changes the user needs to fix.
+    
+  Returns:
+    True if the file requires format changes, False if formatting would produce
+    an identical file.
+  """
+  command_line = CLANG_FORMAT_BASE + CLANG_FORMAT_PARAM_LOG_REPLACEMENTS
+  command_line += filename
+  args = command_line.split(" ")
+  proc = subprocess.Popen(command_line.split(" "), stdout=subprocess.PIPE)
```

Comment:
For Python 3.5+, `subprocess.run` is more convenient and recommended way of running blocking commands. Checkout `capture_output` and `encoding` flags.
https://docs.python.org/3.7/library/subprocess.html